### ISSUE TITLE
return type for bfs

### ISSUE THREAD

**Issue by jitwit** [snowleopard/alga#233]
The current return type/implementation for `bfs` loses connectivity information and makes it unclear whether vertices appear later in the resulting list because they are in a deeper layer of the forest, or because they are in disjoint trees. For example,
```
λ> bfs [1,5] $ circuit [1..3] + circuit [3,2,1] + 5*6
[[1],[2,3],[5],[6]]
it :: [[Int]]
```
Instead, something like 
```
λ> bfs [1,5] $ circuit [1..3] + circuit [3,2,1] + 5*6
[[[1],[2,3]],[[5],[6]]]
it :: [LevelStructure Int]
```
might make more sense, where `type LevelStructure a = [[a]]`?

The change would be:
```
bfs :: [Int] -> AdjacencyIntMap -> [[Int]]
bfs vs = concatMap levels . bfsForestFrom vs
```
to
```
type LevelStructure a = [[a]]

bfs :: [Int] -> AdjacencyIntMap -> [LevelStructure Int]
bfs vs = map levels . bfsForestFrom vs
```

**Comment by snowleopard** [comment]
@jitwit Hmm, shouldn't the result be `[[1,5],[2,3,6]]` instead? That is, vertices 1 and 5 are reachable in 0 steps from the sources, and vertices 2, 3 and 6 are reachable in 1 step?

I can see how your proposed representation preserves even more information but may be somewhat difficult to use (and understand!).

**Comment by jitwit** [comment]
Yeah, there's a lot of nesting with the triple list. I figured it could be useful to keep the components separate, but otherwise I think we can get result you propose with 
```
bfs vs = map concat . Data.List.transpose . bfsForestFrom vs
```

**Comment by snowleopard** [comment]
> I figured it could be useful to keep the components separate

@jitwit But I don't think the components will be separate. Are you referring to strongly-connected components? Then consider the graph `1 * 1 + 1 * 2 + 2 * 2`: it has two connected components `1 * 1` and `2 * 2` but depending on the order of source vertices you will get different results. With `[1, 2]` you'll get what appears to be a single component, and with `[2, 1]` you'll get two. I think the resulting triply-nested list has a rather non-obvious structure.

**Comment by jitwit** [comment]
Ah you're right, that example shows it doesn't make sense to return `[[[a]]]` for the reason I stated! 

Adding the transpose step would still make sense, so that the levels treat the input vertices the same? Calling  `bfs [1,5] $ circuit [1..3] + circuit [3,2,1] + 5*6` to get `[[1,5],[2,3,6]]`  instead of `[[1],[2,3],[5],[6]]` seems more natural.

**Comment by snowleopard** [comment]
> [[1,5],[2,3,6]] instead of [[1],[2,3],[5],[6]] seems more natural.

Yes, I agree! That would be a very useful change.

**Comment by snowleopard** [comment]
Fixed in #234.
